packages/flutter/lib/src/widgets/page_view.dart

```diff
@@ -682,10 +699,20 @@ class _PageViewState extends State<PageView> {
         if (notification.depth == 0 && widget.onPageChanged != null && notification is ScrollUpdateNotification) {
           final PageMetrics metrics = notification.metrics;
           final int currentPage = metrics.page.round();
-          if (currentPage != _lastReportedPage) {
-            _lastReportedPage = currentPage;
+          if (currentPage != _lastOnPageChangedIndex) {
+            _lastOnPageChangedIndex = currentPage;
             widget.onPageChanged(currentPage);
           }
+        } else if (notification.depth == 0 && widget.pageDidChange != null && notification is ScrollEndNotification) {
+          if (!widget.pageSnapping) {
+            return false;
+          }
+          final PageMetrics metrics = notification.metrics;
+          final int currentPage = metrics.page.round();
```

Comment:
If user drag the page in the middle pass the page change threshold and release and immediately tap and hold the page, this will trigger the callback while the scroll offset is still in the middle of two pages. Is this a intended behavior?

Reply:
No, it supposes to trigger exactly when the page has completely changed. Well, seems there is a dilemma.If we use `metrics.page.toInt()`, it sometimes may not trigger because of the double precision problem.
What about this: 
```
final int currentPage = int.parse(metrics.page.toStringAsFixed(15));
```
 Dart doubles are 64-bit floating-point numbers as specified in the IEEE 754 standard.So it has 52 bits of significand, 11 bits of exponent, and 1 sign bit. So the error's value is about (1/2) ^ -52, which makes 15 a safe value as fraction digits.There is a discussion about this.[https://stackoverflow.com/questions/1458633/how-to-deal-with-floating-point-number-precision-in-javascript](url).

If this solution is acceptable, there would be no need to close this function when pageSnapping equal to false because it would have no conflict with `onPageChanged`.So the question below could also be solved.